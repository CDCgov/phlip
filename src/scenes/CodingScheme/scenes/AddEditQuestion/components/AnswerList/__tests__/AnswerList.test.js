import React from 'react'
import { AnswerList } from '../index'
import { shallow } from 'enzyme'

const answers = [
  { text: 'Yes' },
  { text: 'No' }
]

const props = {
  fields: {
    map: cb => answers.map((answer, index) => cb(answers, index, props.fields)),
    swap: jest.fn(),
    push: jest.fn(),
    remove: jest.fn(),
    get: index => answers[index]
  },
  answerType: 3,
  isEdit: false,
  canModify: true,
  name: 'possibleAnswers'
}

const setup = (other = {}) => {
  return shallow(<AnswerList {...props} {...other} />)
}

describe('AddEditQuestion form -- AnswerList', () => {
  test('should render correctly', () => {
    expect(shallow(<AnswerList {...props} />)).toMatchSnapshot()
  })
  
  test('should remove answer choice if the user clicks the trash can icon', () => {
    const spy = jest.spyOn(props.fields, 'remove')
    const wrapper = setup()
    wrapper.find('Field').at(0).prop('handleDelete')()
    expect(spy).toHaveBeenCalledWith(0)
  })
  
  test('should move answer up if the user clicks the up arrow', () => {
    const spy = jest.spyOn(props.fields, 'swap')
    const wrapper = setup()
    wrapper.find('Field').at(1).prop('handleUp')()
    expect(spy).toHaveBeenCalledWith(1, 0)
  })
  
  test('should move answer down if the user clicks the down arrow', () => {
    const spy = jest.spyOn(props.fields, 'swap')
    const wrapper = setup()
    wrapper.find('Field').at(0).prop('handleDown')()
    expect(spy).toHaveBeenCalledWith(0, 1)
  })
  
  test('should change the name of the list of answers if it\'s a category question', () => {
    const wrapper = setup({ answerType: 2 })
    expect(wrapper.find('Field').at(0).prop('label')).toEqual('Category/Tabs')
  })
  
  test('should change the placeholder text if it\'s a category question', () => {
    const wrapper = setup({ answerType: 2 })
    expect(wrapper.find('Field').at(0).prop('placeholder')).toEqual('Add tab')
  })
  
  test('should not show the add more button if the question is a binary question', () => {
    const wrapper = setup({ answerType: 1 })
    expect(wrapper.find('WithTheme(Button)').length).toEqual(0)
  })
  
  test('should show the add more button if the question is not a binary question', () => {
    const wrapper = setup({ answerType: 2 })
    expect(wrapper.find('WithTheme(Button)').length).toEqual(1)
  })
  
  test('should add new field if the user clicks the add more button', () => {
    const spy = jest.spyOn(props.fields, 'push')
    const wrapper = setup({ answerType: 2 })
    wrapper.find('WithTheme(Button)').simulate('click')
    expect(spy).toHaveBeenCalled()
  })
})
