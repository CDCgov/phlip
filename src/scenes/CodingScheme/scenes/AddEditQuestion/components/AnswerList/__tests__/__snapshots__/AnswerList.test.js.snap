// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddEditQuestion form -- AnswerList should render correctly 1`] = `
<FlexGrid
  align="stretch"
  circular={false}
  container={false}
  flex={false}
  justify="stretch"
  padding={0}
  raised={false}
  style={Object {}}
  type="column"
>
  <FlexGrid
    align="stretch"
    circular={false}
    container={false}
    flex={false}
    justify="stretch"
    padding={0}
    raised={false}
    style={Object {}}
    type="column"
  >
    <Field
      answerType={3}
      canModify={true}
      component={[Function]}
      currentValue={
        Object {
          "text": "Yes",
        }
      }
      fields={
        Object {
          "get": [Function],
          "map": [Function],
          "push": [MockFunction],
          "remove": [MockFunction],
          "swap": [MockFunction],
        }
      }
      handleDelete={[Function]}
      handleDown={[Function]}
      handleUp={[Function]}
      index={0}
      isEdit={false}
      label="Answers"
      name="[object Object],[object Object].text"
      placeholder="Add answer"
      type="text"
    />
  </FlexGrid>
  <FlexGrid
    align="stretch"
    circular={false}
    container={false}
    flex={false}
    justify="stretch"
    padding={0}
    raised={false}
    style={Object {}}
    type="column"
  >
    <Field
      answerType={3}
      canModify={true}
      component={[Function]}
      currentValue={
        Object {
          "text": "No",
        }
      }
      fields={
        Object {
          "get": [Function],
          "map": [Function],
          "push": [MockFunction],
          "remove": [MockFunction],
          "swap": [MockFunction],
        }
      }
      handleDelete={[Function]}
      handleDown={[Function]}
      handleUp={[Function]}
      index={1}
      isEdit={false}
      label=""
      name="[object Object],[object Object].text"
      placeholder="Add answer"
      type="text"
    />
  </FlexGrid>
  <WithTheme(Button)
    color="accent"
    onClick={[Function]}
    raised={false}
    style={
      Object {
        "fontWeight": "normal",
        "marginLeft": 32,
      }
    }
    type="button"
    value="Add more"
  />
</FlexGrid>
`;
